deepchem/models/torch_models/fno.py

```diff
@@ -0,0 +1,246 @@
+    This technique was introduced in [1]_
+
+    .. [1] Li, Zongyi, et al. "Fourier neural operator for parametric partial differential equations." arXiv preprint arXiv:2010.08895 (2020).
+    """
+
+    def __init__(self, input_dim, output_dim, modes, width, dims, depth=4):
+        """Initialize the FNO base model.
+
+        Parameters
+        ----------
+        input_dim: int
+            Dimension of input features
+        output_dim: int
+            Dimension of output features
+        modes: int or tuple
+            Number of Fourier modes to keep in spectral convolution
+        width: int
+            Width of the hidden layers
+        dims: int
+            Spatial dimensionality (1, 2, or 3)
+        depth: int, default 4
+            Number of FNO blocks to stack
+        """
+        super().__init__()
+        self.dims = dims
+        self.input_dim = input_dim
+        self.output_dim = output_dim
+        self.width = width
+
+        self.fc0 = nn.Linear(input_dim, width)
+        self.fno_blocks = nn.Sequential(
+            *[FNOBlock(width, modes, dims=dims) for _ in range(depth)])
+        self.fc1 = nn.Linear(width, 128)
+        self.fc2 = nn.Linear(128, output_dim)
+
+    def _ensure_channel_first(self, x):
+        """Ensure input tensor has channels in the correct position.
+
+        Converts between (batch, *spatial_dims, input_dim) and
+        (batch, input_dim, *spatial_dims) formats.
+
+        Parameters
+        ----------
+        x: torch.Tensor
+            Input tensor
+
+        Returns
+        -------
+        torch.Tensor
+            Tensor with channels in position 1: (batch, input_dim, *spatial_dims)
+        """
+        # If x is (batch, *spatial_dims, input_dim) -> permute -> (batch, input_dim, *spatial_dims)
+        if x.shape[-1] == self.input_dim:
+            perm = (0, x.ndim - 1) + tuple(range(1, x.ndim - 1))
+            return x.permute(*perm).contiguous()
+        # If x is already (batch, input_dim, *spatial_dims), leave it as it is.
+        elif x.shape[1] == self.input_dim:
+            return x
+        else:
+            raise ValueError(
+                f"Expected either (batch, input_dim, *spatial_dims) or "
+                f"(batch, *spatial_dims, input_dim), got {tuple(x.shape)}")
+
+    def forward(self, x):
+        """Forward pass through the FNO model.
+
+        Parameters
+        ----------
+        x: torch.Tensor
+            Input tensor of shape (batch, *spatial_dims, input_dim)
+
+        Returns
+        -------
+        torch.Tensor
+            Output tensor of shape (batch, *spatial_dims, output_dim)
+        """
+        if x.ndim != self.dims + 2:
+            raise ValueError(
+                f"Expected tensor with {self.dims + 2} dims (batch, *spatial_dims, input_dim), got {x.ndim}"
+            )
+
+        x = self._ensure_channel_first(x)
+
+        perm_fc0 = (0, *range(2, x.ndim), 1)
+        x = x.permute(*perm_fc0).contiguous()
```

Comment:
Add inline comments here explaining the permutation step and why it is necessary